colossalai/zero/low_level/low_level_strategy.py

```diff
@@ -0,0 +1,561 @@
+        process_group,
+        master_weights,
+        partition_grad,
+        cpu_offload,
+        overlap_communication,
+        reduce_bucket_size,
+        communication_dtype,
+    ):
+        # param_group that current strategy is working on
+        self.param_group = param_group
+        self._dtype = self.param_group["params"][0].dtype
+
+        if process_group is None:  # if process_group is none, convert to default explicitly
+            process_group = dist.group.WORLD
+
+        self.process_group = process_group
+
+        # if process_group is none, will use the default one
+        self._local_rank = dist.get_rank(group=self.process_group)
+        self._world_size = dist.get_world_size(group=self.process_group)
+
+        # master weights copy
+        self._master_weights = master_weights
+
+        self._cpu_offload = cpu_offload
+
+        # stage 2
+        self._partition_grad = partition_grad
+
+        # ParameterStore will manage the tensor buffers used for zero
+        # it will not manage the tensors used by mixed precision training
+        self._param_store = ParameterStore(process_group)
+        self._grad_store = GradientStore(process_group, partition_grad=partition_grad)
+        self._bucket_store = BucketStore(process_group, reduce_bucket_size=reduce_bucket_size)
+
+        # working and master params for mixed precision training
+        group_params = []
+        for param in param_group["params"]:
+            if param.requires_grad:
+                group_params.append(param)
+        master_param_current_rank = self._create_master_param_current_rank(group_params)
+        param_group["params"] = master_param_current_rank
+        self.working_param_group: List[torch.Tensor] = group_params
+        self.master_param_group: List[torch.Tensor] = master_param_current_rank
+
+        # by default this shouldn't be manipulate
+        self.require_grad_sync = True
+
+        # communication params
+        self._overlap_communication = overlap_communication
+        self._communication_dtype = communication_dtype
+
+        # initialize communication stream for
+        # communication-computation overlapping
+        if self._overlap_communication:
+            self._comm_stream = get_accelerator().Stream()
+
+        # reduction hook is only used if overlapping communication
+        # or stage 2 is used
+        # if it is stage 1 without overlapping, no hook will be attached
+        if self._overlap_communication or self._partition_grad:
+            # we iterate over the working params
+            # on each param, we register a hook to its AccumulateGrad object
+            param_group = self.working_param_group
+            for param in param_group:
+                if param.requires_grad:
+
+                    def _grad_handler(grad, param):
+                        # if run with no_sync context, would not sync grad when backward
+                        if self.require_grad_sync:
```

Comment:
The use of `self` here may lead to memory leak

Reply:
fix in #5824